### PR TITLE
Update instances of scenes which have been reimported.

### LINKED ISSUE #14338: Reimporting a 3D mesh (imported as single mesh) doesn't update it in opened scene, must close and reopen
Godot 3.0 master
Windows 10 64 bits
nVidia geforce 940M

Similar to https://github.com/godotengine/godot/issues/10017 (if not the same?)
I have an opened scene which contains an instance of an imported OBJ scene. I modified the scene in Blender, exported it, but in Godot I don't see any difference. I have to close the scene and reopen it to see the difference.

### LINKED ISSUE #38853: Reimporting a 3D model does not get propagated to inherited scenes
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

**Godot version:**

Godot Engine v3.2.1.stable.official

**OS/device including version:**

Arch Linux (rolling release, last updated at the time of submitting).

**Issue description:**

I have a 3D model imported as a scene (The format does not seem to affect, as I've tried all of them with the same behaviour. I'm using gltf2 for the reproduction example). When the model file gets modified on disk (e.g. by exporting again from a 3d modelling tool), godot reimports the asset, but changes are not shown on the inherited scene.

The chanes are only shown if I open the original scene, and then get propagated to all the scenes where this was instanced.

**Steps to reproduce:**

This uses the minimal reproduction project below. The project has four important resources:

- cube.glb, a cube model
- sphere.glb, a sphere model
- mesh.glb, a copy of either mesh.glb or cube.glb
- MeshInherited.tscn, an inherited version of the mesh.glb scene with no modifications, created by double-clicking mesh.glb in the FileSystem tab, selecting "New Inherited" and immediately saving.

The steps to reproduce:

1. Open the project in godot, and open MeshInherited.tscn. Ensure no other scenes are open. 
2. Minimize godot and, using a file manager, delete mesh.glb and duplicate either cube.glb or sphere.glb to a new file. Rename the new copy as mesh.glb.
3. Go back to godot. A reimport will be triggered, but the inherited scene will show the old mesh.
4. Open mesh.glb (double-click in the FileSystem tab, and choose "Open Anyway"). The new mesh will be shown.
5. Go back to the MeshInherited.tscn tab, the mesh is updated.

**Minimal reproduction project:**
[MeshReimportIssue.zip](https://github.com/godotengine/godot/files/4650343/MeshReimportIssue.zip)

### LINKED ISSUE #55372: 3D assets: GLTF files don't autoupdate on reimport, while GLB does
### Godot version

3.3.4 stable, 3.4 stable, 3.4.1 rc1

### System information

Win 64, dedicated Nvidea 765M

### Issue description

Iterating on 3D assets is crucial to any 3D game development workflow.

I have previously exported gltf2 as `*.glb` files from Blender, which worked fine: Whenever I make a change in Blender and export over the previous `*.glb` file, everything in Godot would auto-update as soon as Godot window becomes active.
However now that I tried exporting `*.gltf` files from Blender, I noticed they don't auto-update, even though they auto-reimport just like the `*.glb` files.

Possibly related to https://github.com/godotengine/godot/issues/38853

### Steps to reproduce

1. Open Blender and export the default cube to your Godot project folder as `blender_cube.glb` as well as `blender_cube.gltf`
2. In Godot drag and drop both the `.glb` as well as the `.gltf` file into the scene.
3. Now go back to Blender and make a simple visual change (like extrude a face of the cube), export again as both `blender_cube.glb` as well as `blender_cube.gltf` into your Godot project folder
4. Make Godot the active window and see both files auto reimporting but only the `.glb` file actually update.

In order for the `.gltf` to also update you need to save the Godot scene and reopen it.

### Minimal reproduction project

_No response_

### LINKED ISSUE #55420: Reimport doesn't update instanced scenes in the 3D editor viewport
### Godot version

v4.0.dev.20211117.official [175690957]

### System information

Windows 11, Vulkan, Nvidia Quadro K2000 461.92

### Issue description

Been tweaking auto mesh LOD generation settings, and Reimporting, and I've noticed that after reimport, the Mesh wasn't changed in any way.. then I disabled Mesh LOD totally, reimported, and the mesh still had Auto LODs active

### Steps to reproduce

Import Mesh With Auto LOD Enabled
Disable Auto LOD in Import Settings
Reimport -> LODs still active and unchanged

Note: Restarting the Editor Loads the changes

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by SaracenOne**
This is a PR meant to address the limitation that reimporting a modified scene from something like a GLTF or FBX file will not have these changes immediately reflected in instances of the currently open scenes. This should resolve most issues, the one thing it doesn't yet do is reload resources merely referenced in a reimported scene. This can be addressed by another pass to recursive walk all nodes and properties to correct any stray resources, but an easier and faster method could be to add a means to directly the override the resources in the cache.

Closes #55420
Closes #55372
Closes #14338
Closes #38853
Closes #38853

### CODE REVIEW

**Comment by lyuma** [comment]
Note from yesterday:
Modifying the base glb for an inherited scene while the inherited scene is open crashes

```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.0.alpha.custom_build (7e2586f140a3921c6a87774304b8912c670bc824)
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] HashMap<Variant,List<Pair<Variant const *,Variant>,DefaultAllocator>::Element *,VariantHasher,VariantComparator,3,8>::getptr (S:\repo\godot-pr-pr\core\templates\hash_map.h:317)
[1] HashMap<Variant,List<Pair<Variant const *,Variant>,DefaultAllocator>::Element *,VariantHasher,VariantComparator,3,8>::has (S:\repo\godot-pr-pr\core\templates\hash_map.h:276)
[2] OrderedHashMap<Variant,Variant,VariantHasher,VariantComparator,3,8>::has (S:\repo\godot-pr-pr\core\templates\ordered_hash_map.h:237)
[3] Dictionary::has (S:\repo\godot-pr-pr\core\variant\dictionary.cpp:166)
[4] Object::has_meta (S:\repo\godot-pr-pr\core\object\object.cpp:931)
[5] EditorNode::get_object_icon (S:\repo\godot-pr-pr\editor\editor_node.cpp:4087)
[6] EditorNode::_update_scene_tabs (S:\repo\godot-pr-pr\editor\editor_node.cpp:332)
[7] EditorNode::_notification (S:\repo\godot-pr-pr\editor\editor_node.cpp:567)
[8] EditorNode::_notificationv (S:\repo\godot-pr-pr\editor\editor_node.h:99)
[9] Object::notification (S:\repo\godot-pr-pr\core\object\object.cpp:848)
[10] SceneTree::_notify_group_pause (S:\repo\godot-pr-pr\scene\main\scene_tree.cpp:857)
[11] SceneTree::process (S:\repo\godot-pr-pr\scene\main\scene_tree.cpp:455)
[12] Main::iteration (S:\repo\godot-pr-pr\main\main.cpp:2693)
[13] OS_Windows::run (S:\repo\godot-pr-pr\platform\windows\os_windows.cpp:629)
[14] widechar_main (S:\repo\godot-pr-pr\platform\windows\godot_windows.cpp:168)
[15] _main (S:\repo\godot-pr-pr\platform\windows\godot_windows.cpp:190)
[16] main (S:\repo\godot-pr-pr\platform\windows\godot_windows.cpp:204)
[17] WinMain (S:\repo\godot-pr-pr\platform\windows\godot_windows.cpp:218)
[18] __scrt_common_main_seh (D:\agent\_work\9\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[19] BaseThreadInitThunk
-- END OF BACKTRACE --
================================================================
```

ERROR: Condition "!is_ancestor_of(p_node)" is true. Returning: false
   at: Node::is_editable_instance (scene\main\node.cpp:1861)
WARNING: res://fox_inherited.tscn:3 - ext_resource, invalid UUID: uid://cvnixq8b1uru - using text path instead: res://Fox.glb
     at: ResourceLoaderText::load (scene\resources\resource_format_text.cpp:429)
ERROR: Condition "!is_ancestor_of(p_node)" is true.
   at: Node::set_editable_instance (scene\main\node.cpp:1846)
ERROR: Condition "!owner_valid" is true.
   at: Node::set_owner (scene\main\node.cpp:1497)
ERROR: Condition "!owner_valid" is true.
   at: Node::set_owner (scene\main\node.cpp:1497)
ERROR: Cannot get path of node as it is not in a scene tree.
   at: (scene\main\node.cpp:1588)
ERROR: Condition "!is_inside_tree()" is true. Returning: false
   at: Node::can_process (scene\main\node.cpp:701)
ERROR: Cannot get path of node as it is not in a scene tree.
   at: (scene\main\node.cpp:1588)

**Comment by SaracenOne** [comment]
Still not quite ready yet, but big update on this:
- [x] Fix reimporting instances inside of instances 
- [x] Fix orphaned resource references on scene states
- [x] Fix reimporting instanced inherited scenes
- [x] Add support for preserving groups
- [x] Add support for preserving connections
- [x] Fix updating history
- [x] Fix recaching EditorPlugins
- [x] Add recache request notification for edited scene nodes
- [ ] Fix reimporting on inherited scene when its opened directly
- [ ] Update all resource references on properties
- [ ] Update internal resource references inside ScriptInstances
- [ ] Add EditorSetting feature flag

**Comment by SaracenOne** [comment]
Okay, with some recent changes I'd actually say this is almost ready. I'd quite like to add abitrary property resource remapping too, but at the very least, I'd say this 'ready' now, in that, it does what it says and doesn't seem to crash or throw errors anymore. Although this seemed like a pretty simple feature, it turned out to be a massive pain with loads of special-case bespoke code required to make the reimporting appear as seamless as possible. Lots of bespoke edge-cases which I had to track down and fix, and while I've debugged this pretty intensely already, this will likely still require some meaningful testing. At the very least though, I've added an editor setting feature flag so we can treat it as a more experimental feature if needed.

**Comment by SoapSpangledGames** [comment]
Does this also handle the case where a glTF (or other file type) is made local to the scene (or children made editable), then re-exported from Blender? All the replies I've read mention inheriting a new scene, but not when the model is made local to an existing scene or when the children are made editable.

My workflow is:

1) Drag a model from the FileSystem tab to the Scene tab.
2) Make the model local.
3) Add BoneAttachment3D nodes as needed to the now local Skeleton3D node.
4) Add children to the BoneAttachment3D nodes, changing the transform of the children as needed.
5) Create AnimationTree sequences based on the imported AnimationPlayer.

If I then add a new animation to a model in Blender, and re-export the file, Godot will re-import the file without the new animation. I have to delete the model from the scene, restart Godot, then repeat steps 1-5. Changing scene tabs doesn't help, closing and re-opening the scene doesn't help, restarting Godot without deleting the model from the scene doesn't help.

Having Godot pick up the animation changes automatically would be awesome, and would save me tons of time.

I'm using Godot 4.0 alpha 11.

**Comment by JonqsGames** [comment]
Is their a workaround or something that could be done to make this mergeable ?
Overall experience in Godot 4 is greatly impacted by the mentionned issues.

File: editor/editor_node.cpp [hunk 1]
```diff
@@ -6284,6 +6850,8 @@ EditorNode::EditorNode() {
 	EDITOR_DEF("interface/inspector/default_color_picker_shape", (int32_t)ColorPicker::SHAPE_OKHSL_CIRCLE);
 	EditorSettings::get_singleton()->add_property_hint(PropertyInfo(Variant::INT, "interface/inspector/default_color_picker_shape", PROPERTY_HINT_ENUM, "HSV Rectangle,HSV Rectangle Wheel,VHS Circle,OKHSL Circle", PROPERTY_USAGE_DEFAULT));
 
+	EDITOR_DEF("interface/editor/update_reimported_scenes", true);
```
Comment (fire) [inline editor/editor_node.cpp:6853]: Unclear we we want another option..

**Comment by fire** [comment]
Test plan:

1. Open blender.
2. Put a default monkey in the center
3. Save the scene as a blend
4. Go to this enhancement executable
5. Make a new 3d scene in a new project
6. Instance the blend file path into the scene
7. When you move the monkey in blender and save
8. When you go back to godot it should update.

**Comment by fire** [comment]
# Found a bug.

`node_3d->get_global_transform();` becomes `node_3d->get_relative_transform(node_3d->get_owner()); ` Similar for 2d.

The error case is the origin is different when called getting global transform (universe) versus it's local "global" transform (scene global) in nested scenes.

**Comment by fire** [comment]
Test plan global transforms:

1. Open blender.
2. Put a default monkey in the center
3. Save the scene as a blend
4. Go to this enhancement executable
5. Make a new 3d scene in a new project
6. Instance the blend file path into the scene
7. When you move the monkey in blender and save
8. When you go back to godot it should update.

Variations

1. Go to blender set the monkey to be 10 on the right side
2. Make a child scene and make instance the blend scene
3. Go to the child scene and make the root 10 units to the top
4. Add a main scene
5. Move the Blender scene monkey forward 10 units.

**Comment by fire** [comment]
I patched. So the change was I calculated the transform of the node 2d and node 3d like `new_additive_node_entry.transform_3d = node_3d->get_relative_transform(node_3d->get_parent());` and saved it in the data structure. Notice that it is a local transform. Not a global transform. Also removed the editor flag to disable reloading.

**Comment by fire** [comment]
There's a bug where the nested scene's path is lost with a blend file change.

File: editor/editor_node.cpp [hunk 2]
```diff
@@ -3874,7 +3876,7 @@ Error EditorNode::load_scene(const String &p_scene, bool p_ignore_broken_deps, b
 		Ref<SceneState> state = sdata->get_state();
 		state->set_path(lpath);
 		new_scene->set_scene_inherited_state(state);
-		new_scene->set_scene_file_path(String());
+		new_scene->set_scene_file_path(lpath);
```
Comment (fire) [inline editor/editor_node.cpp:3879]: @godotengine/import Does anyone know what this does? I did this to cover a case where opening a nested scene file loses a reference to the blend file as per the test case when updating.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
The approach is very interesting, although EditorNode is need of a massive clean up after 4.0.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Flavelius** [comment]
Reimporting a .blend file crashes the editor for me (4.0 beta15), but only if the scene is used in any currently open scenes, so i guess this is related to this change. Is this a known issue or unsupported special case with imported .blends?

**Pull request merged**